app/js/translations.js

```diff
@@ -1,28 +1,61 @@
 const i18n = require('i18next')
+const fsBackend = require('i18next-fs-backend')
+const middleware = require('i18next-http-middleware')
 const path = require('path')
 
 module.exports = {
   setup(options = {}) {
     const subdomainLang = options.subdomainLang || {}
     const availableLngs = Object.values(subdomainLang).map(c => c.lngCode)
 
-    i18n.init({
-      resGetPath: path.resolve(__dirname, '../../', 'locales/__lng__.json'),
-      saveMissing: true,
-      resSetPath: path.resolve(
-        __dirname,
-        '../../',
-        'locales/missing-__lng__.json'
-      ),
-      sendMissingTo: 'fallback',
-      fallbackLng: options.defaultLng || 'en',
-      detectLngFromHeaders: true,
-      useCookie: false,
-      preload: availableLngs,
-      supportedLngs: availableLngs
+    i18n
+      .use(fsBackend)
+      .use(middleware.LanguageDetector)
+      .init({
+        backend: {
+          loadPath: path.join(__dirname, '../../locales/__lng__.json')
+        },
+
+        // Detect language set via setLng query string
+        detection: {
+          order: ['querystring'],
+          lookupQuerystring: 'setLng'
+        },
+
+        // Load translation files synchronously: https://www.i18next.com/overview/configuration-options#initimmediate
+        initImmediate: false,
+
+        // We use the legacy v1 JSON format, so configure interpolator to use
+        // underscores instead of curly braces
+        interpolation: {
+          prefix: '__',
+          suffix: '__',
+          unescapeSuffix: 'HTML',
+          // Disable escaping of interpolated values for backwards compatibility.
+          // We escape the value after it's translated in web, so there's no
+          // security risk
+          escapeValue: false,
+          // Disable nesting in interpolated values, preventing user input
+          // injection via another nested value
+          skipOnVariables: true
+        },
+
+        preload: availableLngs,
+        supportedLngs: availableLngs,
+        fallbackLng: options.defaultLng || 'en'
+      })
+
+    // Make custom language detector for Accept-Language header
+    const headerLangDetector = new middleware.LanguageDetector(i18n.services, {
+      order: ['header']
     })
 
     function setLangBasedOnDomainMiddleware(req, res, next) {
+      // setLng query param takes precedence, so if set ignore the subdomain
+      if (req.originalUrl.includes('setLng')) {
+        return next()
```

Comment:
Sorry to bother on `setLng` again. `setLng` was previously triggering the `showUserOtherLng` logic as well. Shall we restore this? (Line 36-37)

Reply:
Hmm, right.

The only place I could find usage of the `setLng` query param was in the smoke tests: https://github.com/overleaf/web-internal/blob/8335d3185075ed4049931c29c37cfa8bbb42ee53/test/smoke/src/SmokeTests.js#L31. I'm not aware of any users relying on this and I failed to find any hits searching the ingress logs for the last few days (stackdriver keeps erroring beyond that).

So I think I would lean towards removing this?

Reply:
My mental model is that `setLng` is an override, so we should just set the language and move on.

Reply:
> So I think I would lean towards removing this?

That's fine by me.

Reply:
Thanks. Could you approve the PR?